### PR TITLE
Add simple ml pipeline for classifier retweets.

### PR DESCRIPTION
**Pull request by IlyaLos**

### CODE REVIEW

File: src/main/scala/IsRetweetClassifier.scala [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    val spark = SparkSession
+      .builder()
+      .appName("Pipeline")
+      .master("yarn")
+      .getOrCreate()
+
+    import spark.sqlContext.implicits._
+
+    val text = spark
+      .sparkContext.wholeTextFiles("hdfs:///user/ilos/tweets/tweets.txt")
```
Comment (ipetrushin-gd) [inline src/main/scala/IsRetweetClassifier.scala]: is there only a single file with tweets? I suppose it should use output of the streaming apps input

JFYI
you could use Spark SQL api instead: `spark.read.textFile`

File: src/main/scala/IsRetweetClassifier.scala [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    val model = pipeline.fit(structuredData)
+
+    val testPrediction = model.transform(test)
+
+    val testResult = testPrediction
+      .map(sample => if (sample.getAs[Int]("label") == sample.getAs[Double]("prediction").toInt) 1 else 0)
+      .reduce(_+_).toFloat / testPrediction.count()
+
+    val testResultRDD = spark.sparkContext.parallelize(Seq(testResult))
+    testResultRDD.saveAsTextFile("hdfs:///user/ilos/tweets/results.txt")
```
Comment (ipetrushin-gd) [inline src/main/scala/IsRetweetClassifier.scala]: application specific and environment dependent variables, properties (e.g. input/output paths) should be externalized into a dedicated configuration file, so the same binary file can run on multiple environments, and there is no need to recompile classes on configuration change. (the idea is similar to logging configuration)

spark application is able to get the externalized configuration parameters passed as command line arguments (`def main(args: Array[String])` or read in a config file that lies on classpath.

so, the action item here is - externalize configuration by using [TypeSafe config lib](https://github.com/lightbend/config) (see [example](https://dzone.com/articles/typesafe-config-features-and-example-usage))

File: src/main/scala/IsRetweetClassifier.scala [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    println("Model accuracy on test data: ")
```
Comment (ipetrushin-gd) [inline src/main/scala/IsRetweetClassifier.scala]: please use Logging instead of sys.out

File: src/main/scala/IsRetweetClassifier.scala [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+      .add(StructField("label", IntegerType, nullable = false))
+
+    val structuredData = spark
+      .createDataFrame(structuredDataRDD, schema)
+      .withColumn("id", monotonically_increasing_id())
+
+    val trainLength = (structuredData.count() * 0.8).toInt
+    val testLength = (structuredData.count() - trainLength).toInt
+
+    val train = structuredData.limit(trainLength)
```
Comment (ipetrushin-gd) [inline src/main/scala/IsRetweetClassifier.scala]: could you clarify please, do you train the model at all?
I see that only `test` data is utilized?

File: src/main/scala/IsRetweetClassifier.scala [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+      .getOrCreate()
+
+    import spark.sqlContext.implicits._
+
+    val text = spark
+      .sparkContext.wholeTextFiles("hdfs:///user/ilos/tweets/tweets.txt")
+      .map { case (filename, content) => content}
+
+    val tweets = text
+      .flatMap(line => line.split("List\\("))
```
Comment (ipetrushin-gd) [inline src/main/scala/IsRetweetClassifier.scala]: > List\\(

what format/encoding the data stored in?
normally, if plain text is the target then JSON or csv/tsv encoding should be used.

the reason is that majority of readers work with JSON or csv/tsv out of the box, so one doesn't need to white custom parsing logic as in this case.

File: src/main/scala/IsRetweetClassifier.scala [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+    val nonEmptyTweets = tweets
+      .map(line => line.stripLineEnd)
+      .filter(line => line.length() > 0)
+
+    val nonEmptyTweetsNormalized = nonEmptyTweets
+      .map(line => line.replaceAll(", true\\)", ", 1"))
+      .map(line => line.replaceAll(", false\\)", ", 0"))
+      .map(line => line.replaceAll(", und, ", ", ??, "))
```
Comment (ipetrushin-gd) [inline src/main/scala/IsRetweetClassifier.scala]: I suppose the three `map` calls can be combined into a single one, because the normalization transformation is a kind of atomic operation. in other words, normalization is not complete as long as all of the replacements took place.

Comment (ipetrushin-gd) [reply src/main/scala/IsRetweetClassifier.scala]: Please white a unit test for the normalization logic (encapsulate the logic into an object/class and test it as a unit)

File: src/main/scala/IsRetweetClassifier.scala [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    val testPrediction = model.transform(test)
+
+    val testResult = testPrediction
+      .map(sample => if (sample.getAs[Int]("label") == sample.getAs[Double]("prediction").toInt) 1 else 0)
+      .reduce(_+_).toFloat / testPrediction.count()
+
+    val testResultRDD = spark.sparkContext.parallelize(Seq(testResult))
+    testResultRDD.saveAsTextFile("hdfs:///user/ilos/tweets/results.txt")
+    println("Model accuracy on test data: ")
+    println((testResult * 100 + 0.5).toInt)
```
Comment (ipetrushin-gd) [inline src/main/scala/IsRetweetClassifier.scala]: could you clarify please, what is `0.5` for ?

File: src/main/scala/IsRetweetClassifier.scala [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{Row, SparkSession}
+import org.apache.spark.ml.feature.{HashingTF, Tokenizer}
+import org.apache.spark.sql.functions._
+
+object IsRetweetClassifier {
+  def main(args: Array[String]) {
+    val spark = SparkSession
+      .builder()
+      .appName("Pipeline")
```
Comment (ipetrushin-gd) [inline src/main/scala/IsRetweetClassifier.scala]: the name could be more descriptive.

File: src/main/scala/IsRetweetClassifier.scala (hunk 4 above)
Comment (IlyaLos) [reply src/main/scala/IsRetweetClassifier.scala]: Yes, it's a bug. In line 'val model = pipeline.fit(structuredData)' I should use train instead of structuredData. So, I just fit my model on both datasets. I will fix it.

File: src/main/scala/IsRetweetClassifier.scala (hunk 1 above)
Comment (IlyaLos) [reply src/main/scala/IsRetweetClassifier.scala]: I merged all files into one using bash. Is it wrong way?

File: src/main/scala/IsRetweetClassifier.scala (hunk 5 above)
Comment (IlyaLos) [reply src/main/scala/IsRetweetClassifier.scala]: I use saveAsTextFiles method of my DStream object for saving tweets to hdfs. And my file has format List(text of tweet, language of tweet, isRetweet flag). And because text of tweet may content any symbols I can't use simple parsing methods in this case. 
It's the best way that I could find.

File: src/main/scala/IsRetweetClassifier.scala (hunk 7 above)
Comment (IlyaLos) [reply src/main/scala/IsRetweetClassifier.scala]: I want to get my precision as integer and just round the number to the nearest integer value. So 99.4 transforms to 99, but 99.6 transforms to 100.

File: src/main/scala/IsRetweetClassifier.scala (hunk 1 above)
Comment (IlyaLos) [reply src/main/scala/IsRetweetClassifier.scala]: Drop RDDs. Now use Spark SQL only.

**Comment by ipetrushin-gd** [comment]
Could you clarify please, why `
e5e94fb | Add simple ml pipeline for classifier retweets.` commit is already in master? I suppose I should have been merged with master upon the PR approval.

File: src/main/scala/SubstringUdfFunctions.scala [hunk 9]
```diff
@@ -0,0 +1,18 @@
+import org.apache.spark.sql.functions.udf
+
+object SubstringUdfFunctions {
```
Comment (ipetrushin-gd) [inline src/main/scala/SubstringUdfFunctions.scala]: That's rather inappropriate way for doing extraction.
Approach with StructType schema was more correct in that sense.

What you should do is:
 - define schema of an Normalized Tweet via a case class
 - let `normalizeTweets()` play its parsing/extraction role, but make it constructing model objects instead of String representation that required to be parsed downstream. 

here is a pseudo code example:
```scala
case class NormalizedTweet(text: String, language: String, ...)

def normalizeTweet(tweetInRawFormat: String) : NormalizedTweet = {
       ...
}

val normalizedTweets = rawTweets.map(normalizeTweets(_))
```

File: src/main/scala/IsRetweetClassifier.scala (hunk 5 above)
Comment (ipetrushin-gd) [reply src/main/scala/IsRetweetClassifier.scala]: please add TODO to change the output format to JSON / CSV.

File: src/main/scala/IsRetweetClassifier.scala (hunk 7 above)
Comment (ipetrushin-gd) [reply src/main/scala/IsRetweetClassifier.scala]: do you have an idea how to rewrite this part of code so it becomes obvious for anyone who read the code what is the purpose of the value?

File: src/main/scala/IsRetweetClassifier.scala [hunk 10]
```diff
@@ -68,17 +56,16 @@ object IsRetweetClassifier {
...
       .map(sample => if (sample.getAs[Int]("label") == sample.getAs[Double]("prediction").toInt) 1 else 0)
       .reduce(_+_).toFloat / testPrediction.count()
 
-    val testResultRDD = spark.sparkContext.parallelize(Seq(testResult))
-    testResultRDD.saveAsTextFile("hdfs:///user/ilos/tweets/results.txt")
-    println("Model accuracy on test data: ")
-    println((testResult * 100 + 0.5).toInt)
+    val testAccuracyAsIntegerNumberOfPercents = (testResult * 100 + 0.5).toInt
+
+    logger.info("Model accuracy on test data: " + testAccuracyAsIntegerNumberOfPercents.toString)
```
Comment (ipetrushin-gd) [inline src/main/scala/IsRetweetClassifier.scala]: JFYI
a more concise alternative:
```scala
logger.info(s"Model accuracy on test data: $testAccuracyAsIntegerNumberOfPercents")
```

**Comment by IlyaLos** [comment]
My fault :( Fixed now

File: src/test/scala/TweetsNormalizationSpec.scala [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+      "Tweet text, ??, 0",
+      "Tweet text, ru, 1",
+      "Tweet text, ru, 0",
+      "false true false und en ru, en, 0"
+    ))
+
+    actual.collect().foreach(println)
+    expected.collect().foreach(println)
+
+    assert(actual.collect().sameElements(expected.collect()))
```
Comment (ipetrushin-gd) [inline src/test/scala/TweetsNormalizationSpec.scala]: you should use assertion API provided by  [spark-fast-tests](https://github.com/ipetrushin-gd/spark-ml-practice/blob/8d551fa3be89852e8f316b62008006677bf448f4/build.sbt#L14) lib (e.g. `assertSmallDataFrameEquality`) instead. 

no need to reinvent the wheel :)

P.S.
I see that `spark-fast-tests` is quite limited and requires a `SparkSessionWrapper`. Let's add TODO for switching to more concise  and powerful [spark-testing-base](https://github.com/holdenk/spark-testing-base)

File: src/test/scala/TweetsNormalizationSpec.scala [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+      "Tweet text, en, 0",
+      "Tweet text, ??, 1",
+      "Tweet text, ??, 0",
+      "Tweet text, ru, 1",
+      "Tweet text, ru, 0",
+      "false true false und en ru, en, 0"
+    ))
+
+    actual.collect().foreach(println)
+    expected.collect().foreach(println)
```
Comment (ipetrushin-gd) [inline src/test/scala/TweetsNormalizationSpec.scala]: always use logging for operational output. 
we are trying to make a production app not a code running in data scientists notebook :) :)

File: src/main/scala/SparkSessionWrapper.scala [hunk 13]
```diff
@@ -0,0 +1,14 @@
+import com.typesafe.config.ConfigFactory
+import org.apache.spark.sql.SparkSession
+
+trait SparkSessionWrapper {
+  lazy val spark: SparkSession = {
+    val config = ConfigFactory.parseResources("application.conf")
```
Comment (ipetrushin-gd) [inline src/main/scala/SparkSessionWrapper.scala]: Please add a TODO comment with a note to upgrade to a compile safe API (e.g. [pureconfig](https://github.com/pureconfig/pureconfig))
The API allows to white less error prone code and make it more concise.

File: src/main/scala/IsRetweetClassifier.scala [hunk 14]
```diff
@@ -1,49 +1,37 @@
...
   def main(args: Array[String]) {
-    val spark = SparkSession
-      .builder()
-      .appName("Pipeline")
-      .master("yarn")
-      .getOrCreate()
+    import spark.implicits._
 
-    import spark.sqlContext.implicits._
+    val config = ConfigFactory.parseResources("application.conf")
```
Comment (ipetrushin-gd) [inline src/main/scala/IsRetweetClassifier.scala]: so, you are initializing config object in two places (here and in `SparkSessionWrapper`).
Any idea how to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)?

File: src/main/scala/TwitterStream.scala [hunk 15]
```diff
@@ -19,6 +19,7 @@ object TwitterStream extends LazyLogging {
       .map(status => status.getText())
 
     statuses.print()
```
Comment (ipetrushin-gd) [inline src/main/scala/TwitterStream.scala]: log log log :)

File: src/main/resources/application.conf [hunk 16]
```diff
@@ -0,0 +1,8 @@
+input {
+  path = "hdfs:///user/ilos/tweets/tweets.txt"
+  localPath = "/Users/ilyalos/Spark/spark-ml-practice/tweets.txt"
```
Comment (ipetrushin-gd) [inline src/main/resources/application.conf]: > localPath
what is it used for?

normally there should just `path` attribute , and localness/remoteness should be defined by value of this attribute. 

i.e. if you need to define both local and remote configuration, you should create different `application.conf` per local/remote environemnt.

File: src/main/scala/TwitterStream.scala [hunk 17]
```diff
@@ -19,6 +19,7 @@ object TwitterStream extends LazyLogging {
...
+    statuses.repartition(1).saveAsTextFiles("/user/ilos/tweets/tweets_for_timestpamp", "")
```
Comment (ipetrushin-gd) [inline src/main/scala/TwitterStream.scala]: please add TODO comment with a note to move path value to externalized config and change output format to CSV or JSON.

**Review by ipetrushin-gd — CHANGES_REQUESTED** [review]

File: src/main/scala/IsRetweetClassifier.scala (hunk 10 above)
Comment (IlyaLos) [reply src/main/scala/IsRetweetClassifier.scala]: Fixed

File: src/test/scala/TweetsNormalizationSpec.scala (hunk 12 above)
Comment (IlyaLos) [reply src/test/scala/TweetsNormalizationSpec.scala]: Fixed

File: src/main/scala/TwitterStream.scala (hunk 15 above)
Comment (IlyaLos) [reply src/main/scala/TwitterStream.scala]: Fixed

File: src/main/scala/SubstringUdfFunctions.scala (hunk 9 above)
Comment (IlyaLos) [reply src/main/scala/SubstringUdfFunctions.scala]: Fixed

File: src/test/scala/TweetsNormalizationSpec.scala (hunk 11 above)
Comment (IlyaLos) [reply src/test/scala/TweetsNormalizationSpec.scala]: Fixed!

File: src/main/scala/IsRetweetClassifier.scala (hunk 14 above)
Comment (IlyaLos) [reply src/main/scala/IsRetweetClassifier.scala]: Fixed

File: src/main/resources/application.conf (hunk 16 above)
Comment (IlyaLos) [reply src/main/resources/application.conf]: Fixed

File: src/main/scala/IsRetweetClassifier.scala (hunk 10 above)
Comment (IlyaLos) [reply src/main/scala/IsRetweetClassifier.scala]: Fixed

File: src/main/scala/SubstringUdfFunctions.scala (hunk 9 above)
Comment (IlyaLos) [reply src/main/scala/SubstringUdfFunctions.scala]: Fixed

File: src/main/scala/TwitterStream.scala [hunk 18]
```diff
@@ -16,9 +16,13 @@ object TwitterStream extends LazyLogging with ConfigurationWrapper {
     val tweets = TwitterUtils.createFilteredStream(ssc, None, Some(filteredQuery))
 
     val statuses = tweets
-      .map(status => status.getText())
+      .map(status => Seq(status.getText(), status.getLang(), status.isRetweet()))
 
-    statuses.foreachRDD { rdd => if (!rdd.isEmpty()) rdd.collect().foreach { element => logger.info(element) } }
+    statuses.foreachRDD {
+      rdd => if (!rdd.isEmpty()) rdd.collect().foreach {
+        element => logger.info(element.mkString(", "))
```
Comment (ipetrushin-gd) [inline src/main/scala/TwitterStream.scala]: in such kind of Kafka->Spark->HDFS architecture you don't want to log every event on a computation node (i.e. worker). 

it would make sense if producers of events were directly connected to Spark without a durable buffer in between, so we  would need to take care of fault tolerance. 

But, in out case there is no point to introduce redundancy (duplicate/replicate data across all the three components), because Kafka already buffers the events durably. as long as a job is recovered within buffer retention period, no data is going to be lost. 

If the logs are useful for debugging purposes, a corresponding logging level should be used in code (logger.debug(...))  and logger xml configuration. on production INFO level should be set in the configuration.

File: src/main/scala/IsRetweetClassifier.scala [hunk 19]
```diff
@@ -0,0 +1,63 @@
+import com.typesafe.scalalogging.LazyLogging
+import org.apache.spark.ml.Pipeline
+import org.apache.spark.ml.classification.LogisticRegression
+import org.apache.spark.ml.feature.{HashingTF, Tokenizer}
+import org.apache.spark.sql.functions._
+import TweetsNormalization.normalizeTweets
+import FeaturesExtractionFromRawTweet.extractFeaturesFromRawTweet
+
+object IsRetweetClassifier extends LazyLogging with SparkSessionWrapper with ConfigurationWrapper {
```
Comment (ipetrushin-gd) [inline src/main/scala/IsRetweetClassifier.scala]: As per your design:
* definition of SparkSessionWrapper:
```scala
trait SparkSessionWrapper extends ConfigurationWrapper 
```

* definition of  IsRetweetClassifier 
```scala
IsRetweetClassifier extends LazyLogging with SparkSessionWrapper with ConfigurationWrapper
```

I see that there is a redundant inheritance here.

Comment (ipetrushin-gd) [reply src/main/scala/IsRetweetClassifier.scala]: Moreover, from design perspective, it looks like misuse of inheritance.

inheritance should serve its purposes: e.g.  encapsulation,  classification **hierarchy**, grouping related concepts in subtrees, generalization.

In order to reuse code without unnecessary inheritance, you may want to use composition instead.
([an example](https://robots.thoughtbot.com/reusable-oo-composition-vs-inheritance)

Comment (ipetrushin-gd) [reply src/main/scala/IsRetweetClassifier.scala]: Additionally, the naming `...Wrapper` suggest that there is decoration of base logic ([decorator pattern](https://code.tutsplus.com/tutorials/design-patterns-the-decorator-pattern--cms-22641)), which is not quite true as I see - hence misleading. 

[scala decorator example](https://nofluffjuststuff.com/blog/daniel_hinojosa/2009/11/decorator_pattern_in_scala)

File: src/main/scala/IsRetweetClassifier.scala [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+
+    val structuredData = nonEmptyTweetsNormalized
+      .map(extractFeaturesFromRawTweet)
+      .withColumn("id", monotonically_increasing_id())
+
+    val trainLength = (structuredData.count() * 0.8).toInt
+    val testLength = (structuredData.count() - trainLength).toInt
+
+    val train = structuredData.limit(trainLength)
+    val test = structuredData.sort($"id".desc).limit(testLength)
```
Comment (ipetrushin-gd) [inline src/main/scala/IsRetweetClassifier.scala]: FYI:
you can use `Dataset#except` API for this kind of things: e.g. 
```scala
val test = structuredData.except(train)
```

File: src/test/scala/TweetsNormalizationSpec.scala [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+      "Tweet text, en, 0",
+      "Tweet text, ??, 1",
+      "Tweet text, ??, 0",
+      "Tweet text, ru, 1",
+      "Tweet text, ru, 0",
+      "false true false und en ru, en, 0"
+    ))
+
+    actual.collect().foreach { element => logger.info(element) }
+    expected.collect().foreach { element => logger.info(element) }
```
Comment (ipetrushin-gd) [inline src/test/scala/TweetsNormalizationSpec.scala]: normally, we should **avoid** logging in tests unless it's operational info (initializing/setup of test environment) - the idea is that a function which is under a test should do logging on its own (with an appropriate logging level).

File: src/main/scala/IsRetweetClassifier.scala (hunk 19 above)
Comment (ipetrushin-gd) [reply src/main/scala/IsRetweetClassifier.scala]: the above comments are refactoring/design improvements.
please spend about 1 hour trying to understand and fix the concerns. 
if it takes more time, just add TODO and let's move on. we can address it later, according to priorities.

File: src/main/scala/IsRetweetClassifier.scala [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+  def main(args: Array[String]) {
+    import spark.implicits._
+
+    val text = spark.read.textFile(config.getString("input.path"))
+    val aggregatedText = text.agg(concat_ws("\n", collect_list("value"))).as[String]
+
+    val tweets = aggregatedText
+      .flatMap(line => line.split("List\\("))
+
+    val nonEmptyTweets = tweets
```
Comment (ipetrushin-gd) [inline src/main/scala/IsRetweetClassifier.scala]: all steps between  `text` and `structuredData ` are parts of a single ETL process, dedicated to data cleansing / preparation. So, the whole process should be tested.

i.e. I'd like to see a test that checks if `structuredData` was correctly created from  `text` / raw input, taking into account how raw data schema parsing and "empty" tweets are handled in this case.

NOTE: such test may require encapsulation of multiple ETL steps into a single function, so it can be tested as a unit.

File: src/main/scala/TwitterStream.scala (hunk 18 above)
Comment (IlyaLos) [reply src/main/scala/TwitterStream.scala]: Good point! I changed log level to debug.

File: src/main/scala/IsRetweetClassifier.scala (hunk 19 above)
Comment (IlyaLos) [reply src/main/scala/IsRetweetClassifier.scala]: Thx for these links! I think I'm understand difference now. 'ConfigurationWrapper' is incorrect name for that class. But with 'SparkSessionWrapper' is not so clear, but I think it's also incorrect name. So I changed this names. 

Also I removed redundant inheritance.

File: src/main/scala/IsRetweetClassifier.scala (hunk 20 above)
Comment (IlyaLos) [reply src/main/scala/IsRetweetClassifier.scala]: Oh! That's great! Thx

Fixed.

File: src/test/scala/TweetsNormalizationSpec.scala (hunk 21 above)
Comment (IlyaLos) [reply src/test/scala/TweetsNormalizationSpec.scala]: I used this logging as debug output and it's useless now. Removed.

File: src/main/scala/TextToStructuredData.scala [hunk 23]
```diff
@@ -0,0 +1,25 @@
...
+    import spark.implicits._
+
+    val aggregatedText = text.agg(concat_ws("\n", collect_list("value"))).as[String]
+
+    val tweets = aggregatedText
+      .flatMap(line => line.split("List\\("))
+
+    val nonEmptyTweets = tweets
+      .map(line => line.stripLineEnd)
+      .filter(line => line.length() > 0)
```
Comment (ipetrushin-gd) [inline src/main/scala/TextToStructuredData.scala]: FYI: here is how to make it more compact yet readable:

```scala
tweets
 .map(_.stripLineEnd)
 .filterNot(_.isEmpty)
```

File: src/test/scala/TextToStructuredDataSpec.scala [hunk 24]
```diff
@@ -0,0 +1,140 @@
...
+import com.typesafe.scalalogging.LazyLogging
+import org.scalatest.FunSpec
+import TextToStructuredData.getStructuredDataFromText
+import org.apache.spark.sql.Row
+
+//TODO: switch to more concise and powerful spark-testing-base
+class TextToStructuredDataSpec extends FunSpec with SparkSessionCreator with LazyLogging with DatasetComparer {
+  import spark.implicits._
+
+  it("Simple general check") {
```
Comment (ipetrushin-gd) [inline src/test/scala/TextToStructuredDataSpec.scala]: Please apply use-case/corner-case descriptive naming to tests. Take a look [here](http://www.scalatest.org/user_guide/selecting_a_style) for a examples.
Ideally, one should be able to reason what kind of features an implementation supports by looking at test cases. So, if the test cases represent use-cases, it becomes clear.

File: src/main/scala/TextToStructuredData.scala [hunk 25]
```diff
@@ -0,0 +1,25 @@
+import org.apache.spark.sql.functions._
+import TweetsNormalization.normalizeTweets
```
Comment (ipetrushin-gd) [inline src/main/scala/TextToStructuredData.scala]: I see that there is no package structure defined for the artifacts. This is not how Java/Scala software developed nowadays. 
Just compare the line to the previous one and you'll see what I'm taking about:
```scala
org.apache.spark.sql.functions._
```
Please learn about how to structure packages and what should be naming for open source and commercial apps. Here is just an example where you can find the info: https://docs.oracle.com/javase/tutorial/java/package/namingpkgs.html

File: src/test/scala/TextToStructuredDataSpec.scala [hunk 26]
```diff
@@ -0,0 +1,140 @@
...
+    val actual = getStructuredDataFromText(source)
+
+    val expected = spark.createDataFrame(spark.sparkContext.makeRDD(List(
+      Row("\nRT @emy_wng: Thanks Bloomberg for featuring @Ripple and #MoneyTap! https://t.co/WDuJnD1tq4","en",1,0.toLong),
+      Row("$btc #bitcoin #cryptolife\n\n#cryptofamily Good morning, Bitcoin has good support now,\nhe might not instantly jump up,…\nhttps://t.co/ok8vMxuE17","en",0,1.toLong),
+      Row("Earn bitcoin on a daily basis!\n1. Follow @slidecoin\n2. Complete instructions in pinned tweet","en",0,2.toLong)
+    )), actual.schema)
+
+    actual.foreach(row => {println(row.toString())})
+    expected.foreach(row => {println(row.toString())})
```
Comment (ipetrushin-gd) [inline src/test/scala/TextToStructuredDataSpec.scala]: please don't do printing to sys out :) 
use logging with debug level if required. 
and I doubt that this particular case requires it at all.

File: src/test/scala/TextToStructuredDataSpec.scala [hunk 27]
```diff
@@ -0,0 +1,140 @@
...
+
+    val expected = spark.createDataFrame(spark.sparkContext.makeRDD(List(
+      Row("RT @emy_wng: Thanks Bloomberg for featuring @Ripple and #MoneyTap! https://t.co/WDuJnD1tq4","??",1,0.toLong)
+    )), actual.schema)
+
+    assertSmallDataFrameEquality(actual, expected)
+  }
+
+  it("Empty tweet check") {
+    val source = spark.createDataset[String](List(
```
Comment (ipetrushin-gd) [inline src/test/scala/TextToStructuredDataSpec.scala]: FYI: you can create DataFrames/Datasets in a more concise way:

```scala
      import spark.implicits._

      val x = List("...", "...").toDF() //.toDS[T]
```

File: src/test/scala/TextToStructuredDataSpec.scala [hunk 28]
```diff
@@ -0,0 +1,140 @@
+import com.github.mrpowers.spark.fast.tests.DatasetComparer
+import com.typesafe.scalalogging.LazyLogging
+import org.scalatest.FunSpec
+import TextToStructuredData.getStructuredDataFromText
+import org.apache.spark.sql.Row
+
+//TODO: switch to more concise and powerful spark-testing-base
+class TextToStructuredDataSpec extends FunSpec with SparkSessionCreator with LazyLogging with DatasetComparer {
```
Comment (ipetrushin-gd) [inline src/test/scala/TextToStructuredDataSpec.scala]: > with LazyLogging

I don't see that it's utilized here...

File: src/test/scala/TweetsNormalizationSpec.scala [hunk 29]
```diff
@@ -0,0 +1,35 @@
+import org.scalatest.FunSpec
+import TweetsNormalization.normalizeTweets
+import com.typesafe.scalalogging.LazyLogging
+import com.github.mrpowers.spark.fast.tests.DatasetComparer
+
+//TODO: switch to more concise and powerful spark-testing-base
+class TweetsNormalizationSpec extends FunSpec with SparkSessionCreator with LazyLogging with DatasetComparer {
```
Comment (ipetrushin-gd) [inline src/test/scala/TweetsNormalizationSpec.scala]: > with LazyLogging

I don't see that it's utilized here...

File: src/main/scala/TextToStructuredData.scala [hunk 30]
```diff
@@ -0,0 +1,25 @@
...
+
+    val nonEmptyTweets = tweets
+      .map(line => line.stripLineEnd)
+      .filter(line => line.length() > 0)
+
+    val nonEmptyTweetsNormalized = nonEmptyTweets.transform[String](normalizeTweets())
+
+    nonEmptyTweetsNormalized
+      .map(extractFeaturesFromRawTweet)
+      .withColumn("id", monotonically_increasing_id())
```
Comment (ipetrushin-gd) [inline src/main/scala/TextToStructuredData.scala]: could you clarify please, is there a real need to the `id`? where is it utilized?

File: src/main/scala/TextToStructuredData.scala [hunk 31]
```diff
@@ -0,0 +1,25 @@
+import org.apache.spark.sql.functions._
+import TweetsNormalization.normalizeTweets
+import FeaturesExtractionFromRawTweet.extractFeaturesFromRawTweet
+import org.apache.spark.sql.{DataFrame, Dataset}
+
+object TextToStructuredData extends SparkSessionCreator {
+  def getStructuredDataFromText(text: Dataset[String]): DataFrame = {
```
Comment (ipetrushin-gd) [inline src/main/scala/TextToStructuredData.scala]: you have defined a data model / schema for a tweet in `StructuredTweet` class, but the API performing the extraction doesn't utilizes it (i.e. returns `DataFrame`, not `StructuredTweet`). 

If the function does something else or more than it's expected to do (just building structured data from plan text) then it takes too much responsibility (i.e. incoherent).
Or it could that the `StructuredTweet` is not defined well enough to represent all required fields, so additional high-level model of `Row` is used here. In this case, `StructuredTweet` should just include the missing fields.

The goal is to make API utilize the defined model, so it's clear for consumers (at compile time) what data schema is.

File: src/main/scala/TextToStructuredData.scala (hunk 23 above)
Comment (IlyaLos) [reply src/main/scala/TextToStructuredData.scala]: Fixed

File: src/test/scala/TextToStructuredDataSpec.scala (hunk 26 above)
Comment (IlyaLos) [reply src/test/scala/TextToStructuredDataSpec.scala]: Just forgot to delete my debug output.

File: src/test/scala/TextToStructuredDataSpec.scala (hunk 27 above)
Comment (IlyaLos) [reply src/test/scala/TextToStructuredDataSpec.scala]: Fixed

File: src/test/scala/TextToStructuredDataSpec.scala (hunk 28 above)
Comment (IlyaLos) [reply src/test/scala/TextToStructuredDataSpec.scala]: Fixed

File: src/test/scala/TweetsNormalizationSpec.scala (hunk 29 above)
Comment (IlyaLos) [reply src/test/scala/TweetsNormalizationSpec.scala]: Fixed

File: src/main/scala/TextToStructuredData.scala (hunk 30 above)
Comment (IlyaLos) [reply src/main/scala/TextToStructuredData.scala]: There are no needs for it. It's just for 'code style' :) Deleted

Comment (ipetrushin-gd) [reply src/main/scala/TextToStructuredData.scala]: So, is `StructuredTweet` sufficient? i.e. no need to use plain `Row` as a model?

File: src/main/scala/TextToStructuredData.scala (hunk 25 above)
Comment (ipetrushin-gd) [reply src/main/scala/TextToStructuredData.scala]: here is related info as well: http://docs.scala-lang.org/style/naming-conventions.html#packages

File: src/test/scala/TextToStructuredDataSpec.scala (hunk 24 above)
Comment (IlyaLos) [reply src/test/scala/TextToStructuredDataSpec.scala]: I tried to fix it.

File: src/main/scala/TextToStructuredData.scala (hunk 30 above)
Comment (IlyaLos) [reply src/main/scala/TextToStructuredData.scala]: I don't understand your question. Could you please clarify what are you asking about?

File: src/main/scala/TextToStructuredData.scala (hunk 25 above)
Comment (IlyaLos) [reply src/main/scala/TextToStructuredData.scala]: Updated.

File: src/main/scala/TextToStructuredData.scala (hunk 31 above)
Comment (IlyaLos) [reply src/main/scala/TextToStructuredData.scala]: Change returned type to Dataset[StructuredData]. This method take Dataset[String] as argument and concat some lines of it at first. So it's cannot be implemented as String to StructuredData method (many to one relationship).
